File: android/app/src/main/kotlin/com/splitfree/core/ExpenseCategory.kt
package com.splitfree.core

/**
 * Expense categories, grouped the way people think about a shared budget.
 *
 * [wireName] matches the iOS app's raw values exactly, because these travel in
 * the `.splitfree` interchange file. Renaming a display title is safe; renaming
 * a wire name breaks files written by the other platform.
 */
enum class ExpenseCategory(val wireName: String, val group: CategoryGroup) {
    GROCERIES("groceries", CategoryGroup.FOOD_AND_DRINK),
    DINING_OUT("diningOut", CategoryGroup.FOOD_AND_DRINK),
    LIQUOR("liquor", CategoryGroup.FOOD_AND_DRINK),
    COFFEE("coffee", CategoryGroup.FOOD_AND_DRINK),
    RENT("rent", CategoryGroup.HOME),
    MORTGAGE("mortgage", CategoryGroup.HOME),
    UTILITIES("utilities", CategoryGroup.HOME),
    ELECTRICITY("electricity", CategoryGroup.HOME),
    WATER("water", CategoryGroup.HOME),
    INTERNET("internet", CategoryGroup.HOME),
    PHONE_BILL("phoneBill", CategoryGroup.HOME),
    HOUSEHOLD_SUPPLIES("householdSupplies", CategoryGroup.HOME),
    FURNITURE("furniture", CategoryGroup.HOME),
    MAINTENANCE("maintenance", CategoryGroup.HOME),
    CLEANING("cleaning", CategoryGroup.HOME),
    FLIGHT("flight", CategoryGroup.TRANSPORT),
    TRAIN("train", CategoryGroup.TRANSPORT),
    BUS("bus", CategoryGroup.TRANSPORT),
    TAXI("taxi", CategoryGroup.TRANSPORT),
    FUEL("fuel", CategoryGroup.TRANSPORT),
    PARKING("parking", CategoryGroup.TRANSPORT),
    CAR_RENTAL("carRental", CategoryGroup.TRANSPORT),
    PUBLIC_TRANSIT("publicTransit", CategoryGroup.TRANSPORT),
    MOVIES("movies", CategoryGroup.ENTERTAINMENT),
    MUSIC("music", CategoryGroup.ENTERTAINMENT),
    GAMES("games", CategoryGroup.ENTERTAINMENT),
    SPORTS("sports", CategoryGroup.ENTERTAINMENT),
    EVENTS("events", CategoryGroup.ENTERTAINMENT),
    MEDICAL("medical", CategoryGroup.LIFE),
    INSURANCE("insurance", CategoryGroup.LIFE),
    EDUCATION("education", CategoryGroup.LIFE),
    CHILDCARE("childcare", CategoryGroup.LIFE),
    GIFTS("gifts", CategoryGroup.LIFE),
    CLOTHING("clothing", CategoryGroup.LIFE),
    PERSONAL_CARE("personalCare", CategoryGroup.LIFE),
    PETS("pets", CategoryGroup.LIFE),
    CHARITY("charity", CategoryGroup.LIFE),
    TAXES("taxes", CategoryGroup.LIFE),
    SUBSCRIPTIONS("subscriptions", CategoryGroup.LIFE),
    HOTEL("hotel", CategoryGroup.TRAVEL),
    ACTIVITIES("activities", CategoryGroup.TRAVEL),
    SOUVENIRS("souvenirs", CategoryGroup.TRAVEL),
    GENERAL("general", CategoryGroup.OTHER),
    ;

    companion object {
        fun fromWire(value: String): ExpenseCategory =
            entries.firstOrNull { it.wireName == value } ?: GENERAL

        /**
         * Best-effort category from an expense title, so someone who types
         * "Uber to airport" and never opens the picker still gets it filed.
         */
        fun suggestion(title: String): ExpenseCategory? {
            val text = title.lowercase()
            if (text.isEmpty()) return null
            return keywordMap.firstOrNull { (keywords, _) -> keywords.any { text.contains(it) } }?.second
        }

        private val keywordMap: List<Pair<List<String>, ExpenseCategory>> = listOf(
            listOf("uber", "lyft", "cab", "taxi", "ola", "grab ride") to TAXI,
            listOf("flight", "airline", "airfare", "boarding", "delta", "united air") to FLIGHT,
            listOf("hotel", "airbnb", "hostel", "motel", "lodge", "resort") to HOTEL,
            listOf("grocer", "supermarket", "safeway", "tesco", "aldi", "kroger", "trader joe", "whole foods") to GROCERIES,
            listOf("coffee", "starbucks", "latte", "espresso", "cafe", "café") to COFFEE,
            listOf("beer", "wine", "bar tab", "cocktail", "pub", "liquor", "brewery") to LIQUOR,
            listOf("dinner", "lunch", "brunch", "breakfast", "restaurant", "pizza", "sushi", "takeout", "doordash", "ubereats") to DINING_OUT,
            listOf("rent") to RENT,
            listOf("mortgage") to MORTGAGE,
            listOf("electric", "power bill") to ELECTRICITY,
            listOf("water bill") to WATER,
            listOf("internet", "wifi", "broadband", "comcast", "xfinity") to INTERNET,
            listOf("phone bill", "mobile bill", "cellular", "verizon", "at&t") to PHONE_BILL,
            listOf("gas station", "petrol", "fuel", "diesel", "shell", "chevron") to FUEL,
            listOf("parking", "garage fee", "meter") to PARKING,
            listOf("train", "rail", "amtrak", "metro card") to TRAIN,
            listOf("bus ticket", "greyhound", "coach") to BUS,
            listOf("car rental", "hertz", "avis", "enterprise rent") to CAR_RENTAL,
            listOf("movie", "cinema", "theater", "theatre") to MOVIES,
            listOf("concert", "festival", "ticket") to EVENTS,
            listOf("game", "playstation", "xbox", "steam", "nintendo") to GAMES,
            listOf("gym", "yoga", "climbing", "ski pass", "lift ticket") to SPORTS,
            listOf("doctor", "pharmacy", "clinic", "hospital", "dentist", "medicine") to MEDICAL,
            listOf("insurance", "premium") to INSURANCE,
            listOf("tuition", "course", "textbook", "school fee") to EDUCATION,
            listOf("daycare", "babysit", "nanny") to CHILDCARE,
            listOf("gift", "present", "birthday") to GIFTS,
            listOf("clothes", "clothing", "shoes", "jacket", "zara", "h&m", "uniqlo") to CLOTHING,
            listOf("haircut", "salon", "spa", "barber") to PERSONAL_CARE,
            listOf("vet", "dog food", "cat food", "petco") to PETS,
            listOf("donation", "charity", "fundraiser") to CHARITY,
            listOf("tax", "irs") to TAXES,
            listOf("netflix", "spotify", "subscription", "hulu", "icloud", "prime") to SUBSCRIPTIONS,
            listOf("cleaning", "cleaner", "housekeep") to CLEANING,
            listOf("furniture", "ikea", "couch", "mattress", "desk") to FURNITURE,
            listOf("repair", "plumber", "electrician", "handyman") to MAINTENANCE,
            listOf("detergent", "paper towel", "toilet paper", "soap", "supplies") to HOUSEHOLD_SUPPLIES,
            listOf("souvenir", "postcard", "magnet") to SOUVENIRS,
            listOf("tour", "museum", "excursion", "hike", "snorkel", "kayak") to ACTIVITIES,
        )
    }
}

enum class CategoryGroup {
    FOOD_AND_DRINK, HOME, TRANSPORT, ENTERTAINMENT, TRAVEL, LIFE, OTHER;

    val categories: List<ExpenseCategory>
        get() = ExpenseCategory.entries.filter { it.group == this }
}

File: SplitFree/Models/ExpenseCategory.swift
import Foundation
import SwiftUI

/// Expense categories, grouped the way people actually think about a shared budget.
/// Stored by `rawValue` so renaming a title never invalidates saved data.
enum ExpenseCategory: String, CaseIterable, Codable, Identifiable, Sendable {
    // Food & drink
    case groceries, diningOut, liquor, coffee
    // Home
    case rent, mortgage, utilities, electricity, water, internet, phoneBill
    case householdSupplies, furniture, maintenance, cleaning
    // Transport
    case flight, train, bus, taxi, fuel, parking, carRental, publicTransit
    // Entertainment
    case movies, music, games, sports, events
    // Life
    case medical, insurance, education, childcare, gifts, clothing, personalCare
    case pets, charity, taxes, subscriptions
    // Travel
    case hotel, activities, souvenirs
    // Fallback
    case general

    var id: String { rawValue }

    var title: String {
        switch self {
        case .groceries: String(localized: "Groceries")
        case .diningOut: String(localized: "Dining out")
        case .liquor: String(localized: "Drinks")
        case .coffee: String(localized: "Coffee")
        case .rent: String(localized: "Rent")
        case .mortgage: String(localized: "Mortgage")
        case .utilities: String(localized: "Utilities")
        case .electricity: String(localized: "Electricity")
        case .water: String(localized: "Water")
        case .internet: String(localized: "Internet")
        case .phoneBill: String(localized: "Phone")
        case .householdSupplies: String(localized: "Household supplies")
        case .furniture: String(localized: "Furniture")
        case .maintenance: String(localized: "Maintenance")
        case .cleaning: String(localized: "Cleaning")
        case .flight: String(localized: "Flight")
        case .train: String(localized: "Train")
        case .bus: String(localized: "Bus")
        case .taxi: String(localized: "Taxi")
        case .fuel: String(localized: "Fuel")
        case .parking: String(localized: "Parking")
        case .carRental: String(localized: "Car rental")
        case .publicTransit: String(localized: "Transit")
        case .movies: String(localized: "Movies")
        case .music: String(localized: "Music")
        case .games: String(localized: "Games")
        case .sports: String(localized: "Sports")
        case .events: String(localized: "Events")
        case .medical: String(localized: "Medical")
        case .insurance: String(localized: "Insurance")
        case .education: String(localized: "Education")
        case .childcare: String(localized: "Childcare")
        case .gifts: String(localized: "Gifts")
        case .clothing: String(localized: "Clothing")
        case .personalCare: String(localized: "Personal care")
        case .pets: String(localized: "Pets")
        case .charity: String(localized: "Charity")
        case .taxes: String(localized: "Taxes")
        case .subscriptions: String(localized: "Subscriptions")
        case .hotel: String(localized: "Lodging")
        case .activities: String(localized: "Activities")
        case .souvenirs: String(localized: "Souvenirs")
        case .general: String(localized: "General")
        }
    }

    var symbol: String {
        switch self {
        case .groceries: "cart.fill"
        case .diningOut: "fork.knife"
        case .liquor: "wineglass.fill"
        case .coffee: "cup.and.saucer.fill"
        case .rent: "house.fill"
        case .mortgage: "building.columns.fill"
        case .utilities: "bolt.horizontal.fill"
        case .electricity: "bolt.fill"
        case .water: "drop.fill"
        case .internet: "wifi"
        case .phoneBill: "phone.fill"
        case .householdSupplies: "shippingbox.fill"
        case .furniture: "sofa.fill"
        case .maintenance: "wrench.and.screwdriver.fill"
        case .cleaning: "sparkles"
        case .flight: "airplane"
        case .train: "tram.fill"
        case .bus: "bus.fill"
        case .taxi: "car.fill"
        case .fuel: "fuelpump.fill"
        case .parking: "parkingsign"
        case .carRental: "car.2.fill"
        case .publicTransit: "lightrail.fill"
        case .movies: "film.fill"
        case .music: "music.note"
        case .games: "gamecontroller.fill"
        case .sports: "figure.run"
        case .events: "ticket.fill"
        case .medical: "cross.case.fill"
        case .insurance: "shield.lefthalf.filled"
        case .education: "graduationcap.fill"
        case .childcare: "figure.and.child.holdinghands"
        case .gifts: "gift.fill"
        case .clothing: "tshirt.fill"
        case .personalCare: "scissors"
        case .pets: "pawprint.fill"
        case .charity: "heart.fill"
        case .taxes: "doc.text.fill"
        case .subscriptions: "arrow.triangle.2.circlepath"
        case .hotel: "bed.double.fill"
        case .activities: "figure.hiking"
        case .souvenirs: "bag.fill"
        case .general: "list.bullet"
        }
    }

    var group: CategoryGroup {
        switch self {
        case .groceries, .diningOut, .liquor, .coffee:
            .foodAndDrink
        case .rent, .mortgage, .utilities, .electricity, .water, .internet,
             .phoneBill, .householdSupplies, .furniture, .maintenance, .cleaning:
            .home
        case .flight, .train, .bus, .taxi, .fuel, .parking, .carRental, .publicTransit:
            .transport
        case .movies, .music, .games, .sports, .events:
            .entertainment
        case .hotel, .activities, .souvenirs:
            .travel
        case .medical, .insurance, .education, .childcare, .gifts, .clothing,
             .personalCare, .pets, .charity, .taxes, .subscriptions:
            .life
        case .general:
            .other
        }
    }

    var color: Color { group.color }

    /// Best-effort category from an expense title — powers the "smart" default
    /// when someone types "Uber to airport" and never opens the category picker.
    static func suggestion(for title: String) -> ExpenseCategory? {
        let text = title.lowercased()
        guard !text.isEmpty else { return nil }
        for (keywords, category) in keywordMap {
            if keywords.contains(where: { text.contains($0) }) { return category }
        }
        return nil
    }

    private static let keywordMap: [([String], ExpenseCategory)] = [
        (["uber", "lyft", "cab", "taxi", "ola", "grab ride"], .taxi),
        (["flight", "airline", "airfare", "boarding", "delta", "united air"], .flight),
        (["hotel", "airbnb", "hostel", "motel", "lodge", "resort"], .hotel),
        (["grocer", "supermarket", "safeway", "tesco", "aldi", "kroger", "trader joe", "whole foods"], .groceries),
        (["coffee", "starbucks", "latte", "espresso", "cafe", "café"], .coffee),
        (["beer", "wine", "bar tab", "cocktail", "pub", "liquor", "brewery"], .liquor),
        (["dinner", "lunch", "brunch", "breakfast", "restaurant", "pizza", "sushi", "takeout", "doordash", "ubereats"], .diningOut),
        (["rent"], .rent),
        (["mortgage"], .mortgage),
        (["electric", "power bill"], .electricity),
        (["water bill"], .water),
        (["internet", "wifi", "broadband", "comcast", "xfinity"], .internet),
        (["phone bill", "mobile bill", "cellular", "verizon", "at&t"], .phoneBill),
        (["gas station", "petrol", "fuel", "diesel", "shell", "chevron"], .fuel),
        (["parking", "garage fee", "meter"], .parking),
        (["train", "rail", "amtrak", "metro card"], .train),
        (["bus ticket", "greyhound", "coach"], .bus),
        (["car rental", "hertz", "avis", "enterprise rent"], .carRental),
        (["movie", "cinema", "theater", "theatre"], .movies),
        (["concert", "festival", "ticket"], .events),
        (["game", "playstation", "xbox", "steam", "nintendo"], .games),
        (["gym", "yoga", "climbing", "ski pass", "lift ticket"], .sports),
        (["doctor", "pharmacy", "clinic", "hospital", "dentist", "medicine"], .medical),
        (["insurance", "premium"], .insurance),
        (["tuition", "course", "textbook", "school fee"], .education),
        (["daycare", "babysit", "nanny"], .childcare),
        (["gift", "present", "birthday"], .gifts),
        (["clothes", "clothing", "shoes", "jacket", "zara", "h&m", "uniqlo"], .clothing),
        (["haircut", "salon", "spa", "barber"], .personalCare),
        (["vet", "dog food", "cat food", "petco"], .pets),
        (["donation", "charity", "fundraiser"], .charity),
        (["tax", "irs"], .taxes),
        (["netflix", "spotify", "subscription", "hulu", "icloud", "prime"], .subscriptions),
        (["cleaning", "cleaner", "housekeep"], .cleaning),
        (["furniture", "ikea", "couch", "mattress", "desk"], .furniture),
        (["repair", "plumber", "electrician", "handyman"], .maintenance),
        (["detergent", "paper towel", "toilet paper", "soap", "supplies"], .householdSupplies),
        (["souvenir", "postcard", "magnet"], .souvenirs),
        (["tour", "museum", "excursion", "hike", "snorkel", "kayak"], .activities),
    ]
}

enum CategoryGroup: String, CaseIterable, Identifiable, Sendable {
    case foodAndDrink, home, transport, entertainment, travel, life, other

    var id: String { rawValue }

    var title: String {
        switch self {
        case .foodAndDrink: String(localized: "Food & drink")
        case .home: String(localized: "Home")
        case .transport: String(localized: "Transport")
        case .entertainment: String(localized: "Entertainment")
        case .travel: String(localized: "Travel")
        case .life: String(localized: "Life")
        case .other: String(localized: "Other")
        }
    }

    var color: Color {
        switch self {
        case .foodAndDrink: Palette.categoryOrange
        case .home: Palette.categoryIndigo
        case .transport: Palette.categoryBlue
        case .entertainment: Palette.categoryPink
        case .travel: Palette.categoryTeal
        case .life: Palette.categoryPurple
        case .other: Palette.categoryGray
        }
    }

    var categories: [ExpenseCategory] {
        ExpenseCategory.allCases.filter { $0.group == self }
    }
}
